{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "68b37aea-7064-43ad-ab5c-775cd8378181",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Data Types, Variables and Functions\n",
    "\n",
    "As we learn about the basics of programming as illustrated in Python, it will be helpful to recognize some of the different data types (kinds of values) that are commonly used in programming. Some of the most important data types in Python are:\n",
    " - Integers\n",
    " - Floating Point Numbers (floats)\n",
    " - Strings\n",
    " - Boolean (logical) values\n",
    "In addition to the simple data types, there are a few data types that are used to group the simple types together. The compound data types that we will encounter most often in Python are:\n",
    "\n",
    "- List\n",
    "- Sequence\n",
    "- Dictionary\n",
    " \n",
    " The data represented by these data types are often stored in *variables*, usually by *assignment*. The data in variables (and other data) is often manipulated by *functions*. Important concepts that frequently arise in the context of functions:\n",
    " \n",
    " - Parameters\n",
    " - Arguments\n",
    "   + Default values\n",
    " - Output (Return Value)\n",
    " - Side Effect\n",
    " \n",
    "# Statements, Functions, Loops and Conditionals\n",
    "\n",
    "The predominant paradigm for programming, called *structured programming*, says that all programming languages will have the following characteristics:\n",
    "\n",
    "  1. The ability to execute instructions in sequence.\n",
    "  2. The ability to follow different instructions based on the evaluation of conditions.\n",
    "  3. The ability to perform a set of instructions repeatedly.\n",
    "  4. The ability to divide up instructions into modules that can be executed as necessary.\n",
    "Note that not every program written in these languages will use all these features, but the language itself should have all these capabilities.\n",
    "\n",
    "In Python, instructions are executed in the order they appear in the program. An instruction normally ends when the line on which the instruction is written ends. There are exceptions to this, but normally each line is an instruction.\n",
    "\n",
    "The most common construct to conditionally execute instructions is the *if* statement. The word \"if\" is followed by a logical expression (that evaluates to True or False), and a set of instructions that are executed if the statement is true. These instructions are grouped together by indenting the conditional instructions. \"If\" statements are optionally followed by \"else\" statements that provide a set of instructions to execute if the logical expression evaluates to False.\n",
    "\n",
    "The most common form of repetition in Python programs involves a *for* statement. The \"for\" keyword is provided with a list of values[^1] that are assigned to a variable. The statement is followed by a group of (indented) instructions that are executed for each value in the list. Groups of statements can also be looped with a *while* statement, which evaluates a logical expression at the beginning of the loop and executes the loop until the expression evaluates to False.\n",
    "\n",
    "Python programs can be divided up by writing functions that are called by other parts of the program. These functions can be a part of the program itself, or they can be written in other files and subsequently made available with an *import* statement. Python comes with many functions that have been organized into *modules*, and other modules and functions have been created by other parties.\n",
    "\n",
    "All four of the components of programming languages are illustrated in the program below. Look for the conditional statements, loops and functions, and see if you can figure out what the different parts of the program do."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "db7dabab-d384-44b3-ae2d-bf70eecfea0b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# This program generates a rainbow using Turtle graphics. The rainbow\n",
    "# isn't a bow; it's more of a prismatic diagram of color.\n",
    "\n",
    "import turtle\n",
    "from time import sleep\n",
    "\n",
    "#Get turtle positioned to color\n",
    "turtle.home()\n",
    "sleep(10)\n",
    "turtle.colormode(255)\n",
    "turtle.penup()\n",
    "turtle.backward(300)\n",
    "turtle.left(90)\n",
    "turtle.pendown()\n",
    "\n",
    "# Define a function for drawing lines while changing a specific color\n",
    "def rampcolor(colordict, channel, increasing=True):\n",
    "    \"\"\"Takes a dictionary of color settings and the string describing one\n",
    "    of the channels, and adjusts that channel through the range of the\n",
    "    channel. Defaults to increasing the channel, but can adjust downward if\n",
    "    the increasing parameter is set to False.\"\"\"\n",
    "    if increasing:\n",
    "        colorlist = range(0,255,5)\n",
    "    else:\n",
    "        colorlist = range(255,0,-5)\n",
    "    for x in colorlist:\n",
    "        colordict[channel] = x\n",
    "        turtle.pencolor((colordict['r'], colordict['g'], colordict['b']))\n",
    "        turtle.forward(100)\n",
    "        if (x % 10 == 0):\n",
    "            turtle.right(90)\n",
    "            turtle.forward(1)\n",
    "            turtle.right(90)\n",
    "        else:\n",
    "            turtle.left(90)\n",
    "            turtle.forward(1)\n",
    "            turtle.left(90)\n",
    "\n",
    "# Create the dictionary of colors. Red is on full, green and blue are off.\n",
    "colors = {'r':255, 'g':0, 'b':0}\n",
    "rampcolor(colors, 'g')\n",
    "rampcolor(colors, 'r', False)\n",
    "rampcolor(colors, 'b')\n",
    "rampcolor(colors, 'g', False)\n",
    "rampcolor(colors, 'r')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c00cd9c-c855-409d-a217-88d784b36b6f",
   "metadata": {},
   "source": [
    "# Object-Oriented Programming\n",
    "\n",
    "A popular enhancement of structured programming is object-oriented programming, which collects related variables and functions together into an *object*. The variables and functions of an object, called *attributes* and *methods*, respectively, are defined as a *class*. Objects are then created as *instances* of the class. This often makes it conceptually easier to perform actions on collections of data.\n",
    "\n",
    "Object-oriented programming is deeply embedded in Python. Many of the main data types (such as strings, lists, and dictionaries) are objects in Python, and have their own associated methods. The standard library also defines several classes for different purposes, and users also have the option of defining their own classes. Attributes and methods in Python are accessed by following the name of the object with a period and the name of the attribute or method. \n",
    "\n",
    "The code below defines a new class of Turtle, and instantiates that class with an object (called franklin). This particular class definition takes advantage of a concept called *inheritance*, which means that franklin can do everything a normal (computer) turtle can do, plus a new activity. Try running the program below, and then change the instructions to make franklin do different things, or even to create another turtle!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e4e0dc4b-dab2-4804-a87b-7a8a729b8968",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import turtle\n",
    "from time import sleep\n",
    "\n",
    "class TurtleGon(turtle.Turtle):\n",
    "    def colorgon(self, colorlist, side=50, fillcolor=None):\n",
    "        angle = 180 - 180 * (len(colorlist) - 2) / len(colorlist)\n",
    "        if fillcolor is not None:\n",
    "            self.fillcolor(fillcolor)\n",
    "            self.begin_fill()\n",
    "        for color in colorlist:\n",
    "            self.pencolor(color)\n",
    "            self.forward(side)\n",
    "            self.left(angle)\n",
    "        if fillcolor is not None:\n",
    "            self.end_fill()\n",
    "        return\n",
    "\n",
    "franklin = TurtleGon()\n",
    "sleep(5)\n",
    "franklin.colorgon([\"red\",\"green\",\"orange\"], fillcolor=\"purple\")\n",
    "franklin.penup()\n",
    "franklin.forward(75)\n",
    "franklin.pendown()\n",
    "franklin.colorgon([\"teal\",\"magenta\",\"gray\",\"yellow\",\"maroon\",\"white\",\"cyan\",\"silver\"],fillcolor=\"gold\")\n",
    "sleep(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "30f9dda6-c06c-43bd-9c33-663ec3929697",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "franklin.clear()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8ea910e-ca81-4b6b-9394-320abd418b6f",
   "metadata": {},
   "source": [
    "# Biopython\n",
    "\n",
    "There have been a large number of modules for Python that have been created for scientific computing. The [SciPy](http://www.scipy.org) project has compiled several different modules for science, math and engineering. Another project of interest, which leverages the functionality of SciPy (NumPy in particular), is [Biopython](http://biopython.org). Biopython is geared toward computational molecular biology. Much of the documentation is geared around sequence analysis, but there are several modules available.\n",
    "\n",
    "The program below searches for primer sequences in a file called sequence.gb, in GenBank format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6835406-2f2f-4900-88f1-0ba782b3aa95",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This program reads in a sequence from a GenBank file and the sequences\n",
    "# of primers (in EMBOSS's primersearch format) used to amplify a region\n",
    "# of the DNA in the sequence file. The program outputs this sequence to\n",
    "# another file in FASTA format. This version will only analyze one sequence\n",
    "# and one primer set, and will output only the largest sequence that would\n",
    "# be amplified by that primer set. The program could easily be modified to\n",
    "# accommodate more than one sequence and more than one set of primers.\n",
    "\n",
    "# This will rely heavily on Biopython's sequence file parsers.\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Create a dictionary with set name, forward primer sequence\n",
    "# and reverse primer sequence.\n",
    "primers = {\"set\":\"rs2472304\", \"forward\":\"CTAGGGCCAGAGAAAGCTAATG\",\n",
    "             \"reverse\":\"GGATGGTGAAGGGCAAGAA\"}\n",
    "\n",
    "# Find the primers in the sequence record. If the forward primer is in\n",
    "# the sequence, create a new sequence that starts at the position where\n",
    "# the forward primer is located. Then reverse the sequence, and find the\n",
    "# position of the reverse primer. Create a sequence of the reversed sequence\n",
    "# that starts at the position of the reverse primer. The reverse of this\n",
    "# truncated sequence is the sequence that will be amplified in PCR, in\n",
    "# the direction of the forward primer.\n",
    "#\n",
    "# If the forward primer is not found in the sequence, the sequence is\n",
    "# calculated as before, starting with the reverse primer. The truncated\n",
    "# sequence is still reported in the direction of the forward primer, though.\n",
    "\n",
    "for fullsequence in SeqIO.parse('sequence.gb', 'genbank'):\n",
    "    forwardpos = fullsequence.seq.find(primers['forward'])\n",
    "    if forwardpos >= 0:\n",
    "        forwardseq = fullsequence[forwardpos:]\n",
    "        reverseseq = forwardseq.reverse_complement()\n",
    "        reversepos = reverseseq.seq.find(primers['reverse'])\n",
    "        if reversepos >= 0:\n",
    "            pcrsequence = reverseseq[reversepos:].reverse_complement()\n",
    "            pcrsequence.id = primers['set']\n",
    "            pcrsequence.description = \"{0} bp amplimer\".format(len(pcrsequence))\n",
    "            print(pcrsequence.format(\"fasta\"))\n",
    "    else:\n",
    "        reversepos = fullsequence.seq.find(primers['reverse'])\n",
    "        if reversepos >= 0:\n",
    "            reverseseq = fullsequence[reversepos:]\n",
    "            forwardseq = reverseseq.reverse_complement()\n",
    "            forwardpos = forwardseq.seq.find(primers['forward'])\n",
    "            if forwardpos >= 0:\n",
    "                pcrsequence = forwardseq[forwardpos:]\n",
    "                pcrsequence.id = primers['set']\n",
    "                pcrsequence.description = \"{0} bp amplimer\".format(len(pcrsequence))\n",
    "                print(pcrsequence.format(\"fasta\"))\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
